### PR TITLE
HSEARCH-3833 Writing on different local-heap indexes will provoke LockObtainFailedException

### PR DESCRIPTION
**Pull request by fax4ever**
https://hibernate.atlassian.net/browse/HSEARCH-3833

### CODE REVIEW

File: integrationtest/backend/lucene/src/test/java/org/hibernate/search/integrationtest/backend/lucene/search/LuceneSearchMultiIndexIT.java [hunk 1]
```diff
@@ -63,9 +71,19 @@
...
+
+	public LuceneSearchMultiIndexIT(String directoryType) {
+		this.directoryType = directoryType;
+	}
+
 	@Before
 	public void setup() {
-		setupHelper.start()
+		setupHelper.start( BACKEND_1_NAME )
+				.withBackendProperty( BACKEND_1_NAME, "directory.type", directoryType )
```
Comment (yrodiere) [inline integrationtest/backend/lucene/src/test/java/org/hibernate/search/integrationtest/backend/lucene/search/LuceneSearchMultiIndexIT.java]: You no longer need to define and pass a backend name. You can call `start()` and `withBackendProperty( "directory.type", directoryType )`.

File: backend/lucene/src/main/java/org/hibernate/search/backend/lucene/lowlevel/directory/impl/LocalFileSystemDirectoryProvider.java [hunk 2]
```diff
@@ -58,7 +58,7 @@ public void initialize(DirectoryProviderInitializationContext context) {
 		this.directoryRoot = ROOT.get( propertySource ).toAbsolutePath();
 		FileSystemAccessStrategyName accessStrategyName = FILESYSTEM_ACCESS_STRATEGY.get( propertySource );
 		this.accessStrategy = FileSystemAccessStrategy.get( accessStrategyName );
-		this.lockFactory = context.createConfiguredLockFactory().orElseGet( FSLockFactory::getDefault );
```
Comment (yrodiere) [inline backend/lucene/src/main/java/org/hibernate/search/backend/lucene/lowlevel/directory/impl/LocalFileSystemDirectoryProvider.java]: Why do you call `get` here? You'll end up having the exact same problem as with the local heap directory, if the user selects the single-instance lock factory explicitly through configuration.

File: backend/lucene/src/main/java/org/hibernate/search/backend/lucene/lowlevel/directory/impl/DirectoryProviderInitializationContextImpl.java [hunk 3]
```diff
@@ -93,16 +94,16 @@ public ConfigurationPropertySource getConfigurationPropertySource() {
 		}
 	}
 
-	private static LockFactory createLockFactory(LockingStrategyName name) {
```
Comment (yrodiere) [inline backend/lucene/src/main/java/org/hibernate/search/backend/lucene/lowlevel/directory/impl/DirectoryProviderInitializationContextImpl.java]: This should probably be renamed to `createLockFactorySupplier`.

File: backend/lucene/src/main/java/org/hibernate/search/backend/lucene/lowlevel/directory/impl/DirectoryProviderInitializationContextImpl.java [hunk 4]
```diff
@@ -69,7 +70,7 @@ public ConfigurationPropertySource getConfigurationPropertySource() {
 	}
 
 	@Override
-	public Optional<LockFactory> createConfiguredLockFactory() {
```
Comment (yrodiere) [inline backend/lucene/src/main/java/org/hibernate/search/backend/lucene/lowlevel/directory/impl/DirectoryProviderInitializationContextImpl.java]: This should probably be renamed to `createConfiguredLockFactorySupplier`.

**Comment by fax4ever** [comment]
Merged thanks!

**Pull request merged**